src/exporter/web_server.cc

```diff
@@ -0,0 +1,283 @@
+    check_deadline(self);
+  }
+
+private:
+  ssl::stream<tcp::socket> socket_;
+
+  void read_request(std::shared_ptr<https_connection> self) {
+    http::async_read(socket_, buffer_, request_,
+                      [self](beast::error_code ec, std::size_t bytes_transferred) {
+                        boost::ignore_unused(bytes_transferred);
+                        if (ec) {
+                            dout(1) << "ERROR: " << ec.message() << dendl;
+                            return;
+                        }
+                        self->process_request();
+                      });
+  }
+
+  void write_response() override {
+    auto self = shared_from_this();
+    response_.prepare_payload();
+    http::async_write(socket_, response_,
+                      [self](beast::error_code ec, std::size_t) {
+                        self->socket_.async_shutdown([self](beast::error_code ec) {
+                          self->deadline_.cancel();
+                          if (ec) {
+                            dout(1) << "ERROR: " << ec.message() << dendl;
+                          }
+                        });
+                      });
+  }
+
+  void close_connection(beast::error_code& ec) override {
+      socket_.lowest_layer().close(ec);
+  }
+
+};
+
+void http_server(tcp::acceptor &acceptor, tcp::socket &socket) {
+  acceptor.async_accept(socket, [&](beast::error_code ec) {
+    if (!ec) {
+      std::make_shared<http_connection>(std::move(socket))->start();
+    }
+    http_server(acceptor, socket);
+  });
+}
+
+void https_server(tcp::acceptor &acceptor, ssl::context &ssl_ctx) {
+  acceptor.async_accept([&](beast::error_code ec, tcp::socket socket) {
+    if (!ec) {
+      std::make_shared<https_connection>(ssl::stream<tcp::socket>(std::move(socket), ssl_ctx))->start();
+    }
+    https_server(acceptor, ssl_ctx);
+  });
+}
+
+void run_http_server(const std::string& exporter_addr, unsigned short port) {
+  net::io_context ioc{1};
+  tcp::acceptor acceptor{ioc, {net::ip::make_address(exporter_addr), port}};
+  tcp::socket socket{ioc};
+
+  http_server(acceptor, socket);
+
+  dout(1) << "HTTP server running on " << exporter_addr << ":" << port << dendl;
+  ioc.run();
+}
+
+void run_https_server(const std::string& exporter_addr, unsigned short port, const std::string& cert_file, const std::string& key_file) {
+  net::io_context ioc{1};
+  ssl::context ssl_ctx(ssl::context::tlsv12);
+
+  ssl_ctx.use_certificate_chain_file(cert_file);
+  ssl_ctx.use_private_key_file(key_file, ssl::context::pem);
+
+  tcp::acceptor acceptor{ioc, {net::ip::make_address(exporter_addr), port}};
+  https_server(acceptor, ssl_ctx);
+
+  dout(1) << "HTTPS server running on " << exporter_addr << ":" << port << dendl;
+  ioc.run();
+}
+
+void web_server_thread_entrypoint() {
+  try {
+    std::string exporter_addr = g_conf().get_val<std::string>("exporter_addr");
+    unsigned short port = g_conf().get_val<int64_t>("exporter_http_port");
```

Comment:
TCP ports can't have more than 16 bits (unsigned)

Reply:
changed

Reply:
reverted.... There is a compilation warning if we use just a "short" type:

```
web_server.cc: In function ‘void run_http_server(const std::string&, short int)’:
    /home/jolmomar/Code/ceph/src/exporter/web_server.cc:232:70: warning: narrowing conversion of ‘port’ from ‘short int’ to ‘boost::asio::ip::port_type’ {aka ‘short unsigned int’} [-Wnarrowing]
      232 |   tcp::acceptor acceptor{ioc, {net::ip::make_address(exporter_addr), port}};
```